### PR TITLE
[Merged by Bors] - feat(Order/SupClosed): Add lemmas for inserting upper or lower bounds into a Sup/Inf closed set

### PR DESCRIPTION
**Pull request by mans0954**
Add lemmas for inserting upper or lower bounds into a Sup/Inf closed set.

e.g. When `s` is sup closed, so is `s` with `⊤` or `⊥` appended. Similarly for `s` inf closed.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/SupClosed.lean [hunk 1]
```diff
@@ -143,6 +153,16 @@ lemma infClosed_pi {ι : Type*} {α : ι → Type*} [∀ i, SemilatticeInf (α i
     {t : ∀ i, Set (α i)} (ht : ∀ i ∈ s, InfClosed (t i)) : InfClosed (s.pi t) :=
   fun _a ha _b hb _i hi ↦ ht _ hi (ha _ hi) (hb _ hi)
 
+lemma InfClosed.insert_upper_bound {s : Set α} {a : α} (hs : InfClosed s) (ha : ∀ b ∈ s, b ≤ a) :
+    InfClosed (insert a s):= by
```
Comment (YaelDillies) [inline Mathlib/Order/SupClosed.lean:157]:
```suggestion
    InfClosed (insert a s) := by
```

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer delegate

**Comment by mans0954** [comment]
I thought I was going to need this in #18705, but I found a better way. This PR may still be of independent interest though?

**Comment by YaelDillies** [comment]
Oh, my maintainer delegate didn't go through.

maintainer merge

File: Mathlib/Order/SupClosed.lean [hunk 2]
```diff
@@ -79,6 +79,16 @@ lemma supClosed_pi {ι : Type*} {α : ι → Type*} [∀ i, SemilatticeSup (α i
     {t : ∀ i, Set (α i)} (ht : ∀ i ∈ s, SupClosed (t i)) : SupClosed (s.pi t) :=
   fun _a ha _b hb _i hi ↦ ht _ hi (ha _ hi) (hb _ hi)
 
+lemma SupClosed.insert_upper_bound {s : Set α} {a : α} (hs : SupClosed s) (ha : ∀ b ∈ s, b ≤ a) :
```
Comment (jcommelin) [inline Mathlib/Order/SupClosed.lean:82]: Should `ha` be phrased in terms of`upperBounds s`? Idem dito for the statements below.

Comment (mans0954) [reply Mathlib/Order/SupClosed.lean]: Yes -done thanks. Curiously `aesop` needed a hint one way around but not the other.

File: Mathlib/Order/SupClosed.lean [hunk 3]
```diff
@@ -79,6 +79,17 @@ lemma supClosed_pi {ι : Type*} {α : ι → Type*} [∀ i, SemilatticeSup (α i
     {t : ∀ i, Set (α i)} (ht : ∀ i ∈ s, SupClosed (t i)) : SupClosed (s.pi t) :=
   fun _a ha _b hb _i hi ↦ ht _ hi (ha _ hi) (hb _ hi)
 
+lemma SupClosed.insert_upper_bound {s : Set α} {a : α} (hs : SupClosed s) (ha : a ∈ upperBounds s) :
```
Comment (jcommelin) [inline Mathlib/Order/SupClosed.lean:82]:
```suggestion
lemma SupClosed.insert_upperBounds {s : Set α} {a : α} (hs : SupClosed s) (ha : a ∈ upperBounds s) :
```
and similarly elsewhere

bors d+

**Pull request closed without merging**